File: stream-chat-android-docs/src/main/java/io/getstream/chat/docs/java/client/cms/AndroidIntroduction.java
package io.getstream.chat.docs.java.client.cms;

import android.content.Context;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.getstream.chat.android.client.ChatClient;
import io.getstream.chat.android.client.api.models.QueryChannelsRequest;
import io.getstream.chat.android.client.api.state.ChatClientExtensions;
import io.getstream.chat.android.client.channel.ChannelClient;
import io.getstream.chat.android.client.channel.state.ChannelState;
import io.getstream.chat.android.client.extensions.FlowExtensions;
import io.getstream.chat.android.client.logger.ChatLogLevel;
import io.getstream.chat.android.models.Channel;
import io.getstream.chat.android.models.FilterObject;
import io.getstream.chat.android.models.Filters;
import io.getstream.chat.android.models.Message;
import io.getstream.chat.android.models.UploadAttachmentsNetworkType;
import io.getstream.chat.android.models.User;
import io.getstream.chat.android.models.querysort.QuerySortByField;
import io.getstream.chat.android.models.querysort.QuerySorter;
import io.getstream.chat.android.offline.plugin.factory.StreamOfflinePluginFactory;
import kotlinx.coroutines.flow.StateFlow;

public class AndroidIntroduction {

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=java#chat-client">Chat Client</a>
     */
    public void chatClient(Context applicationContext) {
        String apiKey = "{{ api_key }}";
        String token = "{{ chat_user_token }}";
        // Step 1 - Set up the OfflinePlugin for offline storage
        StreamOfflinePluginFactory offlinePluginFactory = new StreamOfflinePluginFactory(applicationContext);

        // Step 2 - Set up the client, together with offline plugin, for API calls
        ChatClient client = new ChatClient.Builder(apiKey, applicationContext)
                // Change log level
                .logLevel(ChatLogLevel.ALL)
                .withPlugins(offlinePluginFactory)
                .uploadAttachmentsNetworkType(UploadAttachmentsNetworkType.NOT_ROAMING)
                .build();

        // Step 3 - Authenticate and connect the user
        User user = new User.Builder()
                .withId("summer-brook-2")
                .withName("Paranoid Android")
                .withImage("https://bit.ly/2TIt8NR")
                .build();

        client.connectUser(user, token).enqueue((result) -> {
            if (result.isSuccess()) {
                // Handle success
            } else {
                // Handler error
            }
        });
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=java#channels">Channels</a>
     */
    public void channels(ChatClient client, LifecycleOwner lifecycleOwner) {
        ChannelClient channelClient = client.channel("messaging", "travel");

        Map<String, Object> extraData = new HashMap<>();
        extraData.put("name", "Awesome channel about traveling");
        List<String> memberIds = new ArrayList<String>();

        // Creating a channel with the low level client
        channelClient.create(memberIds, extraData).enqueue((result) -> {
            if (result.isSuccess()) {
                Channel channel = result.getOrNull();
                // Use channel by calling methods on channelClient
            } else {
                // Handle error
            }
        });

        // Watching a channel's state using the offline library
        StateFlow<ChannelState> channelStateFlow = ChatClientExtensions.watchChannelAsState(client, "messaging:travel", 0);
        LiveData<ChannelState> channelStateLiveData = FlowExtensions.asLiveData(channelStateFlow);

        channelStateLiveData.observe(lifecycleOwner, channelState -> {
            if (channelState != null) {
                // StateFlow objects to observe. Use FlowExtensions.asLiveData(stateFlow); to LiveData conversion.
                channelState.getMessages();
                channelState.getReads();
                channelState.getTyping();
            } else {
                // User not connected yet.
            }
        });
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=java#messages">Messages</a>
     */
    public void messages(ChannelClient channelClient) {
        Map<String, Object> extraData = new HashMap<>();
        extraData.put("customField", "123");

        Message message = new Message.Builder()
                .withText("I’m mowing the air Rand, I’m mowing the air.")
                .withCid("messaging:travel")
                .withExtraData(extraData)
                .build();


        channelClient.sendMessage(message).enqueue((result) -> {
            if (result.isSuccess()) {
                Message sentMessage = result.getOrNull();
            } else {
                // Handle error
            }
        });
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=java#querying-channels">Querying Channels</a>
     */
    public void queryingChannels(ChatClient client) {
        FilterObject filter = Filters.and(Filters.eq("type", "messaging"), Filters.in("members", "john"));
        QuerySorter<Channel> sort = QuerySortByField.descByName("lastMessageAt");

        int offset = 0;
        int limit = 10;
        int messageLimit = 0;
        int memberLimit = 0;
        QueryChannelsRequest request = new QueryChannelsRequest(filter, offset, limit, sort, messageLimit, memberLimit).withWatch().withState();

        client.queryChannels(request).enqueue((result) -> {
            if (result.isSuccess()) {
                List<Channel> channels = result.getOrNull();
            } else {
                // Handle error
            }
        });
    }
}

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/client/cms/AndroidIntroduction.kt
package io.getstream.chat.docs.kotlin.client.cms

import android.content.Context
import io.getstream.chat.android.client.ChatClient
import io.getstream.chat.android.client.api.models.QueryChannelsRequest
import io.getstream.chat.android.client.api.state.watchChannelAsState
import io.getstream.chat.android.client.channel.ChannelClient
import io.getstream.chat.android.client.logger.ChatLogLevel
import io.getstream.chat.android.models.Channel
import io.getstream.chat.android.models.Filters
import io.getstream.chat.android.models.Message
import io.getstream.chat.android.models.UploadAttachmentsNetworkType
import io.getstream.chat.android.models.User
import io.getstream.chat.android.models.querysort.QuerySortByField
import io.getstream.chat.android.models.querysort.QuerySortByField.Companion.descByName
import io.getstream.chat.android.offline.plugin.factory.StreamOfflinePluginFactory
import io.getstream.result.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

class AndroidIntroduction {

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=kotlin#chat-client">Chat Client</a>
     */
    fun chatClient(applicationContext: Context) {
        val apiKey = "{{ api_key }}"
        val token = "{{ chat_user_token }}"
        // Step 1 - Set up the OfflinePlugin for offline storage
        val offlinePluginFactory = StreamOfflinePluginFactory(
            appContext = applicationContext,
        )

        // Step 2 - Set up the client, together with offline plugin, for API calls
        val client = ChatClient.Builder(apiKey, applicationContext)
            // Change log level
            .logLevel(ChatLogLevel.ALL)
            .withPlugins(offlinePluginFactory)
            .uploadAttachmentsNetworkType(UploadAttachmentsNetworkType.NOT_ROAMING)
            .build()

        // Step 3 - Authenticate and connect the user
        val user = User(
            id = "summer-brook-2",
            name = "Paranoid Android",
            image = "https://bit.ly/2TIt8NR",
        )
        client.connectUser(
            user = user,
            token = token, // or client.devToken(userId); if auth is disabled for your app
        ).enqueue { result ->
            when (result) {
                is Result.Success -> {
                    // Handle success
                }
                is Result.Failure -> {
                    // Handler error
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=kotlin#channels">Channels</a>
     */
    fun channels(client: ChatClient, scope: CoroutineScope) {
        val channelClient = client.channel(channelType = "messaging", channelId = "travel")

        val extraData = mutableMapOf<String, Any>(
            "name" to "Awesome channel about traveling"
        )

        // Creating a channel with the low level client
        channelClient.create(memberIds = emptyList(), extraData = extraData).enqueue { result ->
            when (result) {
                is Result.Success -> {
                    val channel: Channel = result.value
                    // Use channel by calling methods on channelClient
                }
                is Result.Failure -> {
                    // Handler error
                }
            }
        }

        // Watching a channel's state using the offline library
        scope.launch {
            client.watchChannelAsState(cid = "messaging:travel", messageLimit = 0).collect { channelState ->
                if (channelState != null) {
                    // StateFlow objects to observe
                    channelState.messages
                    channelState.reads
                    channelState.typing
                } else {
                    // User not connected yet.
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=kotlin#messages">Messages</a>
     */
    fun messages(channelClient: ChannelClient) {
        val message = Message(
            text = "I’m mowing the air Rand, I’m mowing the air.",
            cid = "messaging:travel",
            extraData = mutableMapOf("customField" to "123")
        )

        channelClient.sendMessage(message).enqueue { result ->
            when (result) {
                is Result.Success -> {
                    val message: Message = result.value
                }
                is Result.Failure -> {
                    // Handler error
                }
            }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/?language=kotlin#querying-channels">Querying Channels</a>
     */
    fun queryingChannels(client: ChatClient) {
        val filter = Filters.and(
            Filters.eq("type", "messaging"),
            Filters.`in`("members", "john"),
        )
        val sort = QuerySortByField<Channel>().descByName("lastMessageAt")

        val request = QueryChannelsRequest(
            filter = filter,
            offset = 0,
            limit = 10,
            querySort = sort
        ).withWatch().withState()

        client.queryChannels(request).enqueue { result ->
            when (result) {
                is Result.Success -> {
                    val channels: List<Channel> = result.value
                }
                is Result.Failure -> {
                    // Handler error
                }
            }
        }
    }
}
